ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -0,0 +1,205 @@
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;list;create;update
+// +kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=roles,verbs=get;list;watch;create;delete;update
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=get;list;watch;create;delete
+// Reconcile used to bridge the desired state with the current state
+func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("rayjob", request.NamespacedName)
+	log.Info("reconciling RayJob", "service NamespacedName", request.NamespacedName)
+
+	// Get serving cluster instance
+	var err error
+	var RayJobInstance *rayv1alpha1.RayJob
+
+	if RayJobInstance, err = r.getRayJobInstance(ctx, request); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var rayClusterInstance *rayv1alpha1.RayCluster
+	if rayClusterInstance, err = r.getOrCreateRayClusterInstance(ctx, RayJobInstance); err != nil {
+		err = r.updateState(ctx, RayJobInstance, rayv1alpha1.FailToGetOrCreateRayCluster, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Info("Updated RayCluster")
+
+	RayJobInstance.Status.RayClusterStatus = rayClusterInstance.Status
+
+	var clientURL string
+	if clientURL, err = r.fetchDashboardURL(ctx, rayClusterInstance); err != nil || clientURL == "" {
+		err = r.updateState(ctx, RayJobInstance, rayv1alpha1.WaitForDashboard, err)
+		return ctrl.Result{}, err
+	}
+
+	rayDashboardClient := utils.RayDashboardClient{}
+	rayDashboardClient.InitClient(clientURL)
+
+	shouldUpdate := r.checkIfNeedSubmitServeDeployment(RayJobInstance, request)
+
+	if shouldUpdate {
+		if err = r.updateServeDeployment(RayJobInstance, rayDashboardClient, request); err != nil {
+			err = r.updateState(ctx, RayJobInstance, rayv1alpha1.FailServeDeploy, err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	var isHealthy bool
+	if isHealthy, err = r.getAndCheckServeStatus(RayJobInstance, rayDashboardClient); err != nil {
+		err = r.updateState(ctx, RayJobInstance, rayv1alpha1.FailGetServeDeploymentStatus, err)
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Check serve health", "isHealthy", isHealthy)
+
+	if isHealthy {
+		RayJobInstance.Status.ServiceStatus = rayv1alpha1.Running
+		if err := r.Status().Update(ctx, RayJobInstance); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		RayJobInstance.Status.RayClusterName = ""
```

Comment:
This part is not needed. You will reuse the same cluster. What you need is to terminate the job and start the job again in the same cluster. I do not think RayJob should terminate a cluster